typed-arrow-dyn/src/projection.rs

```diff
@@ -0,0 +1,495 @@
+//! Projection masks for nested Arrow schemas.
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use arrow_schema::{DataType, Field, FieldRef, Fields, Schema, SchemaRef};
+use parquet::arrow::{ArrowSchemaConverter, ProjectionMask as ParquetProjectionMask};
+use thiserror::Error;
+
+/// Errors surfaced while validating projection paths against an Arrow schema.
+#[derive(Debug, Error)]
+pub enum ProjectionMaskError {
+    /// Projection mask referenced an invalid schema path.
+    #[error("invalid projection path {path:?}: {message}")]
+    InvalidPath {
+        /// Path components that failed validation.
+        path: Vec<usize>,
+        /// Human-readable reason for the failure.
+        message: String,
+    },
+}
+
+/// A set of columns within a potentially nested schema to project. If `None`, all columns
+/// are included.
+///
+/// Paths are expressed as index sequences traversing the schema tree. For example,
+/// `[3, 0, 1]` targets top-level column `3`, then the list element at `0`, and finally
+/// field `1` within the struct stored inside the list.
+///
+/// # Example
+///
+/// ```text
+/// // Schema layout (paths are ProjectionMask indices; `[idx]` is the flattened Parquet leaf):
+/// //
+/// // 0 ─ id: Int64                              [0]
+/// // 1 ─ name: Utf8                             [1]
+/// // 2 ─ address: Struct
+/// //     ├─ 0 ─ street: Utf8                    [2]
+/// //     ├─ 1 ─ city: Utf8                      [3]
+/// //     └─ 2 ─ zip: Utf8                       [4]
+/// // 3 ─ phones: List<Struct>
+/// //     └─ 0 ─ item: Struct
+/// //         ├─ 0 ─ kind: Utf8                  [5]
+/// //         └─ 1 ─ number: Utf8                [6]
+/// ```
+///
+/// Selecting `[0]`, `[2, 1]`, and `[3, 0, 1]` keeps the `id`, `address.city`, and
+/// `phones.item.number` fields. Calling [`ProjectionMask::to_parquet`] then marks Parquet leaves
+/// `0`, `3`, and `6` as `true`, producing the bitmap `[true, false, false, true, false, false,
+/// true]`.
+///
+/// Call [`ProjectionMask::validate`] before using a mask to ensure every indexed path exists in the
+/// runtime schema. Valid masks only clone full top-level fields, keeping downstream
+/// [`DynProjection`](crate::view::DynProjection) construction safe.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ProjectionMask {
```

Comment:
Neither `to_parquet` nor `to_schema` call `validate` or return a `Result`. If a caller forgets to validate, an out-of-bounds index either panics in `to_schema` (lines 117-138) or is silently ignored in `to_parquet` (lines 80-113) producing a mask that selects nothing. The rest of `typed-arrow-dyn` goes out of its way to return structured `DynViewErrors`, so this divergent behavior feels out of place and violates the “no panics on dynamic input” design point in `AGENTS.md`. Please funnel both helpers through validate (propagating `ProjectionMaskError`) so that bad user input can’t crash or mis-project.